### DISCUSSION TITLE
Can only set "Storage Session Cookies" in header of returned response from current UI route Loader?

### DISCUSSION THREAD

**Discussion by ConnorVO** [remix-run/remix#2244] [Q&A]
I have the world's most basic remix app setup to try to figure out createStorageSessionCookies and it seems like they can only be set in the loader of the UI route that is calling the API.

Example:
- Index UI route
- API at routes/api/testApi.tsx (this is a resource route with only an action)
- app/session.server.ts where I create the storage session cookie using createStorageSessionCookies()

If I try to set the cookie anywhere else except for the return from loader in the Index UI route, it doesn't work. If I call the testApi and try to return a response that sets the cookie, doesn't work. If I create a separate function that is called from the loader and that function returns the result of an API call and tries to set the cookie, doesn't work. It literally only works if I set the cookie in the response of the loader.

Example Index route where it works

```
export const loader: LoaderFunction = async ({ request }) => {
    const session = await sessionStorage.getSession(
        request.headers.get('Cookie')
    );
    session.set('cookie_example', 'thisworks');
    return json(
        { status: 'this works' },
        {
            headers: {
                'Set-Cookie': await sessionStorage.commitSession(session),
            },
        }
    );
};

export default function Index() {
    const data = useLoaderData();

    return <div>hi</div>;
}
```

This won't work
```

const setCookie = async (request: Request) => {
    const session = await sessionStorage.getSession(
        request.headers.get('Cookie')
    );
    session.set('cookie_example', 'thisdoesntwork');
    return json(
        { status: 'this doesnt work' },
        {
            headers: {
                'Set-Cookie': await sessionStorage.commitSession(session),
            },
        }
    );
}

export const loader: LoaderFunction = async ({ request }) => {
    const failFunc = await setCookie(request)
    return failFunc.json()
};

export default function Index() {
    const data = useLoaderData();

    return <div>hi</div>;
}

```

Also won't work trying to set them from an API route (app/routes/api/setCookie.tsx called from the loader or from a button - neither work)
```

export let loader: LoaderFunction = () => redirect('/');

export let action: ActionFunction = async ({ request }) => {
    const session = await sessionStorage.getSession(
        request.headers.get('Cookie')
    );
    session.set('cookie_example', 'does not work');

    return json(
        { status: 'does not work' },
        {
            headers: {
                'Set-Cookie': await sessionStorage.commitSession(session),
            },
        }
    );
};
```

**Comment by kiliman** [comment]
How are you creating your session cookie? Do you have `path: '/'` in the config? Otherwise cookies will only be available in the path that sets the cookie.

**Reply by kiliman** [reply]
I didn't see the incorrect code in your second loader example.

Anyway, here's an example that works where cookies are persisted even across redirects.

https://codesandbox.io/s/remix-session-cookies-2hjcij?file=/app/routes/api/post.ts

```ts
// index.tsx
export const loader: LoaderFunction = async ({ request }) => {
  const session = await getSession(request.headers.get('cookie'))
  return json({ message: session.get('message') })
}

export default function Index() {
  const data = useLoaderData()

  return (
    <div>
      <h1>Remix Session Cookies</h1>
      <Form action="/api/post" method="post">
        <input type="text" name="message" defaultValue="hello world" />
        <button>Submit</button>
      </Form>
      <pre>{JSON.stringify(data, null, 2)}</pre>
    </div>
  )
}
```
```ts
// api/post.ts
export const action: ActionFunction = async ({ request }) => {
  const session = await getSession(request.headers.get('cookie'))
  const formData = await request.formData()

  session.set('message', formData.get('message'))
  return redirect('/', {
    headers: { 'set-cookie': await commitSession(session) },
  })
}
```

**Reply by kiliman** [reply]
BTW: I wonder if Remix could optimize the double-GET after redirect.

If you redirect from a POST, Remix will GET that route. Then during re-validation, it will call the loaders of all matching routes (which also includes the one we just redirected to). With nested routing, this could potentially be main duplicate requests.

Remix should probably store which routes were just fetched from the redirect, then remove those from the revalidation set.

**Reply by kiliman** [reply]
One more thing. If you just copied the sample code from the Remix docs, it included a `domain: 'remix.run'` key. You should remove or change that. Otherwise you won't be able to read any of your cookies.

**Comment by sergiodxa** [comment]
There's a problem here.
```ts
export const loader: LoaderFunction = async ({ request }) => {
    const failFunc = await setCookie(request)
    return failFunc.json()
};
```

That will not work, because you are returning the body of the response returned by setCookie without the headers where you do Set-Cookie
